pkg/tlsx/openssl/openssl.go

```diff
@@ -0,0 +1,238 @@
+		if err != nil {
+			return nil, errors.Wrap(err, "could not read ca certificate")
+		}
+		caStore := opensslCtx.GetCertificateStore()
+		err = caStore.LoadCertificatesFromPEM(caCert)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not add certificate to store")
+		}
+	}
+
+	conn, err := openssl.Dial("tcp", address, opensslCtx, c.OpenSSLDialFlags())
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	var resolvedIP string
+	if !iputil.IsIP(hostname) {
+		remoteAddr := conn.RemoteAddr().String()
+		if IP, _, err := net.SplitHostPort(remoteAddr); err == nil {
+			resolvedIP = IP
+		}
+		if resolvedIP == "" {
+			resolvedIP = ip
+		}
+	}
+
+	if options.SNI != "" {
+		if err := conn.SetTlsExtHostName(options.SNI); err != nil {
+			return nil, errors.New("could not set custom SNI")
+		}
+	}
+
+	if err := conn.Handshake(); err != nil {
+		return nil, errors.Wrap(err, "could not do handshake")
+	}
+
+	peerCertificates, err := conn.PeerCertificateChain()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get peer certificates")
+	}
+
+	if len(peerCertificates) == 0 {
+		return nil, errors.New("no certificates returned by server")
+	}
+
+	tlsCipher, err := conn.CurrentCipher()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get current cipher")
+	}
+
+	leafCertificate := peerCertificates[0]
+	certificateChain := peerCertificates[1:]
+	serverName := conn.GetServername()
+
+	x509LeafCertificate, err := c.convertOpenSSLToX509Certificate(leafCertificate)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not convert openssl leaf certificate")
+	}
+
+	now := time.Now()
+	response := &clients.Response{
+		Timestamp:           &now,
+		Host:                hostname,
+		IP:                  resolvedIP,
+		ProbeStatus:         true,
+		Port:                port,
+		Cipher:              tlsCipher,
+		TLSConnection:       "openssl",
+		CertificateResponse: c.convertCertificateToResponse(hostname, x509LeafCertificate),
+		ServerName:          serverName,
+	}
+	if c.options.TLSChain {
+		for _, opensslCert := range certificateChain {
+			x509Cert, err := c.convertOpenSSLToX509Certificate(opensslCert)
+			if err != nil {
+				return nil, errors.Wrap(err, "could not convert openssl chain certificate")
+			}
+			response.Chain = append(response.Chain, c.convertCertificateToResponse(hostname, x509Cert))
+		}
+	}
+	return response, nil
+}
+
+func (c *Client) convertOpenSSLToX509Certificate(opensslCert *openssl.Certificate) (*x509.Certificate, error) {
+	pemBytes, err := opensslCert.MarshalPEM()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not marshal openssl to pem x509")
+	}
+	pemBlock, _ := pem.Decode(pemBytes)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not read openssl pem x509 to go pem")
+	}
+	if pemBlock.Type != "CERTIFICATE" {
+		return nil, errors.Wrap(err, "unsupported pem block type")
+	}
+	x509Certificate, err := x509.ParseCertificate(pemBlock.Bytes)
```

Comment:
we should either use zmap/zcrypto/x509 for decoding openssl certs or use the library implementation itself since in some of the errored cases, go stdlib implementation failed to parse the certificate but the openssl library was able to.

Getting the certificate using openssl and parsing it again from PEM in go removes that advantage